### PR TITLE
LibWeb/CSS: Implement the color-scheme CSS property

### PR DESCRIPTION
**Pull request by Gingeh**
First, some beauty shots:
image
image

Now that you've been thoroughly dazzled, I need to mention a few issues with this implementation:
- System color keywords and the `light-dark` function aren't computed when cascading or when serialized.
- Styles aren't correctly invalidated when the preferred color scheme changes.
- ~~The `theme-color` and `color-scheme` meta tags are only evaluated on insertion.~~ (resolved)

I believe the first two are fundamental problems with our style computation system, it can't handle cases where the computed value of a property is meaningfully distinct from the specified value. It isn't enough to just serialize things differently, the computed values are also needed for style inheritance and invalidation.

### CODE REVIEW

File: Libraries/LibWeb/CSS/Default.css [hunk 1]
```diff
@@ -95,7 +87,7 @@ input[type=range]::-webkit-slider-runnable-track {
     height: 4px;
     width: 100%;
     margin-top: 6px;
-    background-color: Background;
+    background-color: AccentColorText;
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/CSS/Default.css:90]: Should this be `AccentColor`? Using the text colour for non-text feels odd. Same for `progress::-webkit-progress-bar` below.

File: Libraries/LibWeb/CSS/StyleValues/ColorSchemeStyleValue.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    StringBuilder builder;
+    bool first = true;
+    for (auto const& scheme : schemes()) {
+        if (first) {
+            first = false;
+        } else {
+            builder.append(' ');
+        }
+        builder.append(scheme);
+    }
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/CSS/StyleValues/ColorSchemeStyleValue.cpp:26]: I was going to say you could use `builder.join(' ', schemes());` here... but actually this needs to make use of `serialize_an_identifier()` I think, because a scheme could be a `<custom-ident>`. That function escapes any funky characters.

File: Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 3]
```diff
@@ -1129,8 +1129,7 @@ void HTMLInputElement::computed_properties_changed()
     if (!appearance.has_value() || *appearance == CSS::Appearance::None)
         return;
 
-    auto palette = document().page().palette();
-    auto accent_color = palette.color(ColorRole::Accent).to_string();
+    auto accent_color = CSS::CSSKeywordValue::create(CSS::Keyword::Accentcolor)->to_string(Web::CSS::CSSStyleValue::SerializationMode::Normal);
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/HTML/HTMLInputElement.cpp:1132]: This is equivalent to `auto accent_color = "accentcolor";`, is that intentional? If so, the simpler way of doing that is `CSS::string_from_keyword(CSS::Keyword::Accentcolor)`.

File: Libraries/LibWeb/HTML/HTMLProgressElement.cpp [hunk 4]
```diff
@@ -132,8 +133,7 @@ void HTMLProgressElement::update_progress_value_element()
 
 void HTMLProgressElement::computed_properties_changed()
 {
-    auto palette = document().page().palette();
-    auto accent_color = palette.color(ColorRole::Accent).to_string();
+    auto accent_color = CSS::CSSKeywordValue::create(CSS::Keyword::Accentcolor)->to_string(Web::CSS::CSSStyleValue::SerializationMode::Normal);
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/HTML/HTMLProgressElement.cpp:136]: Same comment about `string_from_keyword()`.

File: Libraries/LibWeb/DOM/Document.cpp [hunk 5]
```diff
@@ -1432,6 +1432,17 @@ void Document::set_visited_link_color(Color color)
...
 
+Optional<Vector<String> const&> Document::supported_color_schemes() const
+{
+    return m_supported_color_schemes;
+}
+
+void Document::set_supported_color_schemes(Vector<String> color_schemes)
+{
+    if (!m_supported_color_schemes.has_value())
+        m_supported_color_schemes = color_schemes;
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/DOM/Document.cpp:1443]: I think I get why this only sets once (because the algorithm in `HTMLMetaElement::inserted()` wants to only use the first matching `<meta>`'s value) but it's not quite correct: If the chosen `<meta name="color-scheme">` is removed, modified, or is no longer the first one, then we continue to use its value instead of the new first one (or nothing).

IMO, this setter should always set, and that code in `HTMLMetaElement::inserted()` should be moved into a `Document::obtain_supported_color_schemes()` method or similar, which `HTMLMetaElement::inserted()` just calls. I think this is also true for the `theme-color` code that was already there.

File: Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 6]
```diff
@@ -3491,6 +3492,39 @@ RefPtr<CSSStyleValue> Parser::parse_color_function(TokenStream<ComponentValue>&
...
+
+    auto light = parse_color_value(inner_tokens);
+    if (!light)
+        return {};
+    inner_tokens.discard_whitespace();
+
+    if (!inner_tokens.consume_a_token().is(Token::Type::Comma))
+        return {};
+
+    auto dark = parse_color_value(inner_tokens);
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/CSS/Parser/Parser.cpp:3516]: I think technically `parse_color_value()` already discards leading whitespace, but I'd feel better if you explicitly did so here too. :sweat_smile:

File: Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 7]
```diff
@@ -3491,6 +3492,39 @@ RefPtr<CSSStyleValue> Parser::parse_color_function(TokenStream<ComponentValue>&
...
+    if (!dark)
+        return {};
+    inner_tokens.discard_whitespace();
+
+    if (inner_tokens.has_next_token())
+        return {};
+
+    transaction.commit();
+    return CSSLightDark::create(light, dark);
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/CSS/Parser/Parser.cpp:3525]:
```suggestion
    return CSSLightDark::create(light.release_nonnull(), dark.release_nonnull());
```
(See comments on `CSSLightDark.h`)

File: Libraries/LibWeb/CSS/StyleValues/CSSLightDark.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+private:
+    CSSLightDark(RefPtr<CSSStyleValue> light, RefPtr<CSSStyleValue> dark)
+        : CSSColorValue(CSSColorValue::ColorType::LightDark)
+        , m_properties { .light = move(light), .dark = move(dark) }
+    {
+    }
+
+    struct Properties {
+        RefPtr<CSSStyleValue> light;
+        RefPtr<CSSStyleValue> dark;
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/CSS/StyleValues/CSSLightDark.h:36]: These should be nonnull because we know they always have a value. In fact I think for this struct they also want to be `ValueComparingNonnullRefPtr` so that the automatic `operator==` works properly. The constructor and `create()` methods also need to take them as nonnull. I can't remember if the `ValueComparing...` part is necessary there or it'll happen implicitly.

File: Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    auto const& other_color = other.as_color();
+    if (color_type() != other_color.color_type())
+        return false;
+    auto const& other_light_dark = verify_cast<CSSLightDark>(other_color);
+    return m_properties == other_light_dark.m_properties;
+}
+
+String CSSLightDark::to_string(SerializationMode mode) const
+{
+    // FIXME: We don't have enough information to determine the computed value here.
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp:33]: Is this an actual problem, or has the `light-dark()` function been replaced with an actual color value by then? (Genuine question; this stuff is very confusing. :sweat_smile:)

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Very cool! I really like the parsing code, which probably means I need to get out more. :laughing:

File: Libraries/LibWeb/CSS/Default.css (hunk 1 above)
Comment (Gingeh) [reply Libraries/LibWeb/CSS/Default.css:90]: I'm using `AccentColorText` here because I can rely on it having enough contrast with the `AccentColor` progress bar.

File: Libraries/LibWeb/CSS/StyleValues/ColorSchemeStyleValue.cpp (hunk 2 above)
Comment (Gingeh) [reply Libraries/LibWeb/CSS/StyleValues/ColorSchemeStyleValue.cpp:26]: Ah, forgot about that, done :+1:

File: Libraries/LibWeb/HTML/HTMLInputElement.cpp (hunk 3 above)
Comment (Gingeh) [reply Libraries/LibWeb/HTML/HTMLInputElement.cpp:1132]: Done :+1:

File: Libraries/LibWeb/CSS/Parser/Parser.cpp (hunk 6 above)
Comment (Gingeh) [reply Libraries/LibWeb/CSS/Parser/Parser.cpp:3516]: Fair enough, done :+1:

File: Libraries/LibWeb/CSS/StyleValues/CSSLightDark.h (hunk 8 above)
Comment (Gingeh) [reply Libraries/LibWeb/CSS/StyleValues/CSSLightDark.h:36]: Makes sense, done :+1:

File: Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp (hunk 9 above)
Comment (Gingeh) [reply Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp:33]: I don't think the `light-dark` style value is actually "replaced" at any point, this is relevant to my point in the PR description that our style computation system can't reasonably handle this.

File: Libraries/LibWeb/DOM/Document.cpp (hunk 5 above)
Comment (Gingeh) [reply Libraries/LibWeb/DOM/Document.cpp:1443]: Yeah, a complete implementation would also need to recalculate the supported color schemes on removal and on attribute changes.

Comment (Gingeh) [reply Libraries/LibWeb/DOM/Document.cpp]: Done, although it doesn't actually have much useful observable behaviour since the styles aren't invalidated correctly (or at all).

File: Libraries/LibWeb/CSS/Default.css (hunk 1 above)
Comment (AtkinsSJ) [reply Libraries/LibWeb/CSS/Default.css:90]: Ah gotcha, I missed that.

File: Libraries/LibWeb/DOM/Document.cpp [hunk 10]
```diff
@@ -1432,6 +1435,87 @@ void Document::set_visited_link_color(Color color)
...
+                auto color = css_value->to_color(root_node);
+                document().page().client().page_did_change_theme_color(color);
+                return TraversalDecision::Break;
+            }
+        }
+
+        return TraversalDecision::Continue;
+    });
+
+    // 3. Return nothing(the page has no theme color).
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/DOM/Document.cpp:1516]: I think this also needs to clear the theme color at the beginning?

Also the theme-color change could be its own commit.

File: Libraries/LibWeb/HTML/HTMLMetaElement.cpp [hunk 11]
```diff
@@ -160,4 +136,30 @@ void HTMLMetaElement::inserted()
...
+{
+    Base::removed_from(parent);
+
+    if (name().has_value()) {
+        if (name()->equals_ignoring_ascii_case("theme-color"sv)) {
+            document().obtain_theme_color();
+        } else if (name()->equals_ignoring_ascii_case("color-scheme"sv)) {
+            document().obtain_supported_color_schemes();
+        }
+    }
```
Comment (AtkinsSJ) [inline Libraries/LibWeb/HTML/HTMLMetaElement.cpp:149]: Feels like this repeated bit of code should be pulled out into its own method.

File: Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp (hunk 9 above)
Comment (AtkinsSJ) [reply Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp:33]: I just tried this PR out with http://wpt.live/css/css-color/light-dark-inheritance.html and we currently fail it. From that it seems we're supposed to resolve `light-dark()` before inheriting it. That's related to what I mean: the computed value can't be `light-dark()` itself, but should be whatever `light-dark()` has been resolved to.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
We don't currently pass all of them but it might be good to import some of the light-dark WPT tests as they cover some edge cases:
- http://wpt.live/css/css-color/light-dark-currentcolor-in-color.html
- http://wpt.live/css/css-color/light-dark-inheritance.html
- http://wpt.live/css/css-color/light-dark-currentcolor.html
- http://wpt.live/css/css-color/light-dark-basic.html

File: Libraries/LibWeb/DOM/Document.cpp (hunk 10 above)
Comment (Gingeh) [reply Libraries/LibWeb/DOM/Document.cpp:1516]: Sure. Although I don't know what the default color is and can't test it because it's only used by the AppKit UI.

File: Libraries/LibWeb/HTML/HTMLMetaElement.cpp (hunk 11 above)
Comment (Gingeh) [reply Libraries/LibWeb/HTML/HTMLMetaElement.cpp:149]: Okay, done :+1:

File: Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp (hunk 9 above)
Comment (Gingeh) [reply Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp:33]: Yes, that's what should happen, unfortunately (as I have said) I am unable to find any way to make the style computer actually compute the style value. This is the same problem we had with `PositionStyleValue` which was "solved" with the `SerializationMode` workaround, that simply isn't enough here.

Comment (AtkinsSJ) [reply Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp:33]: Gotcha, I was a bit confused about what you meant exactly. That's a tricky one. :thinking:

Comment (AtkinsSJ) [reply Libraries/LibWeb/CSS/StyleValues/CSSLightDark.cpp:33]: Nah, it's good as-is!

**Review by AtkinsSJ — APPROVED** [review]
